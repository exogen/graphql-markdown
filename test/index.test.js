const fs = require('fs')
const path = require('path')
const tempy = require('tempy')
const {
  run,
  loadSchemaJSON,
  renderSchema,
  updateSchema
} = require('../src/index')

function createPrinter() {
  const printer = chunk => {
    printer.output += `${chunk}\n`
  }
  printer.output = ''
  return printer
}

function createMockConsole() {
  return {
    log: createPrinter(),
    error: createPrinter()
  }
}

function readFile(filename) {
  filename = path.resolve(__dirname, filename)
  return new Promise((resolve, reject) => {
    fs.readFile(filename, 'utf8', (err, data) => {
      err ? reject(err) : resolve(data)
    })
  })
}

describe('renderSchema()', () => {
  it('renders the GraphBrainz schema', async () => {
    const schema = await loadSchemaJSON('graphbrainz/schema.json')
    const printer = createPrinter()
    renderSchema(schema, { printer })
    const expected = await readFile('./fixtures/graphbrainz.md')
    expect(printer.output).toBe(expected)
  })
})

describe('diffSchema()', () => {
  const schemaDiff = require('./fixtures/cover-art-archive')

  it('outputs a JSON schema containing only additions', async () => {
    expect(await schemaDiff).toMatchSnapshot()
  })

  it('can be rendered', async () => {
    const printer = createPrinter()
    renderSchema(await schemaDiff, { printer })
    const expected = await readFile('./fixtures/cover-art-archive.md')
    expect(printer.output).toBe(expected)
  })
})

describe('loadSchemaJSON()', () => {
  it('can load exported GraphQLSchema objects', async () => {
    const moduleSchema = await loadSchemaJSON('graphbrainz/lib/schema')
    expect(moduleSchema.__schema.queryType.name).toBe('Query')
  })

  it('can load JSON schema files', async () => {
    const jsonFileSchema = await loadSchemaJSON('graphbrainz/schema.json')
    expect(jsonFileSchema.__schema.queryType.name).toBe('Query')
  })

  it('can load GraphQL schema language files', async () => {
    const graphqlFileSchema = await loadSchemaJSON(
      path.resolve(__dirname, './fixtures/graphbrainz.graphql')
    )
    expect(graphqlFileSchema.__schema.queryType.name).toBe('Query')
  })
})

describe('updateSchema()', () => {
  const filename = tempy.file()

  it('adds schema and comment markers to file if it does not exist', async () => {
    const {
      initial,
      graphql
    } = await require('./fixtures/graphbrainz-updateSchema')

    await updateSchema(filename, initial, { graphql })
    const actual = await readFile(filename)
    const expected = await readFile(
      './fixtures/graphbrainz-updateSchema-initial.md'
    )
    expect(actual).toBe(expected)
  })

  it('updates schema between comment markers in an existing file', async () => {
    const {
      updated,
      graphql
    } = await require('./fixtures/graphbrainz-updateSchema')

    await updateSchema(filename, updated, { graphql })
    const actual = await readFile(filename)
    const expected = await readFile(
      './fixtures/graphbrainz-updateSchema-updated.md'
    )
    expect(actual).toBe(expected)
  })
})

describe('run()', () => {
  describe('with no args', () => {
    it('prints the help message', () => {
      const console = createMockConsole()
      run([], { console, exit: false })
      expect({
        stdout: console.log.output,
        stderr: console.error.output
      }).toMatchSnapshot()
    })
  })

  describe('with --help arg', () => {
    it('prints the help message', () => {
      const console = createMockConsole()
      run(['--help'], { console, exit: false })
      expect({
        stdout: console.log.output,
        stderr: console.error.output
      }).toMatchSnapshot()
    })
  })
})
