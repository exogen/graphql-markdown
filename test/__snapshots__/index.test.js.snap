// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diffSchema() outputs a JSON schema containing only additions 1`] = `
Object {
  "__schema": Object {
    "directives": Array [],
    "types": Array [
      Object {
        "description": "A [release](https://musicbrainz.org/doc/Release) represents the
unique release (i.e. issuing) of a product on a specific date with specific
release information such as the country, label, barcode, packaging, etc. If you
walk into a store and purchase an album or single, they’re each represented in
MusicBrainz as one release.",
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "An object containing a list and summary of the cover art images that are
present for this release from the [Cover Art Archive](https://musicbrainz.org/doc/Cover_Art_Archive).
This field is provided by the Cover Art Archive extension.",
            "isDeprecated": false,
            "name": "coverArtArchive",
            "type": Object {
              "kind": "OBJECT",
              "name": "CoverArtArchiveRelease",
              "ofType": null,
            },
          },
        ],
        "inputFields": null,
        "interfaces": Array [
          Object {
            "kind": "INTERFACE",
            "name": "Node",
            "ofType": null,
          },
          Object {
            "kind": "INTERFACE",
            "name": "Entity",
            "ofType": null,
          },
        ],
        "kind": "OBJECT",
        "name": "Release",
        "possibleTypes": null,
      },
      Object {
        "description": "A [release group](https://musicbrainz.org/doc/Release_Group) is
used to group several different releases into a single logical entity. Every
release belongs to one, and only one release group.

Both release groups and releases are “albums” in a general sense, but with an
important difference: a release is something you can buy as media such as a CD
or a vinyl record, while a release group embraces the overall concept of an
album – it doesn’t matter how many CDs or editions/versions it had.",
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "The cover art for a release in the release group, obtained from the
[Cover Art Archive](https://musicbrainz.org/doc/Cover_Art_Archive). A
release in the release group will be chosen as representative of the release
group.
This field is provided by the Cover Art Archive extension.",
            "isDeprecated": false,
            "name": "coverArtArchive",
            "type": Object {
              "kind": "OBJECT",
              "name": "CoverArtArchiveRelease",
              "ofType": null,
            },
          },
        ],
        "inputFields": null,
        "interfaces": Array [
          Object {
            "kind": "INTERFACE",
            "name": "Node",
            "ofType": null,
          },
          Object {
            "kind": "INTERFACE",
            "name": "Entity",
            "ofType": null,
          },
        ],
        "kind": "OBJECT",
        "name": "ReleaseGroup",
        "possibleTypes": null,
      },
      Object {
        "description": "An object containing a list of the cover art images for a release obtained
from the [Cover Art Archive](https://musicbrainz.org/doc/Cover_Art_Archive),
as well as a summary of what artwork is available.",
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [
              Object {
                "defaultValue": "FULL",
                "description": "The size of the image to retrieve. By default, the returned image will
have its full original dimensions, but certain thumbnail sizes may be
retrieved as well.",
                "name": "size",
                "type": Object {
                  "kind": "ENUM",
                  "name": "CoverArtArchiveImageSize",
                  "ofType": null,
                },
              },
            ],
            "deprecationReason": null,
            "description": "The URL of an image depicting the album cover or “main front” of the release,
i.e. the front of the packaging of the audio recording (or in the case of a
digital release, the image associated with it in a digital media store).

In the MusicBrainz schema, this field is a Boolean value indicating the
presence of a front image, whereas here the value is the URL for the image
itself if one exists. You can check for null if you just want to determine
the presence of an image.",
            "isDeprecated": false,
            "name": "front",
            "type": Object {
              "kind": "SCALAR",
              "name": "URLString",
              "ofType": null,
            },
          },
          Object {
            "args": Array [
              Object {
                "defaultValue": "FULL",
                "description": "The size of the image to retrieve. By default, the returned image will
have its full original dimensions, but certain thumbnail sizes may be
retrieved as well.",
                "name": "size",
                "type": Object {
                  "kind": "ENUM",
                  "name": "CoverArtArchiveImageSize",
                  "ofType": null,
                },
              },
            ],
            "deprecationReason": null,
            "description": "The URL of an image depicting the “main back” of the release, i.e. the back
of the packaging of the audio recording.

In the MusicBrainz schema, this field is a Boolean value indicating the
presence of a back image, whereas here the value is the URL for the image
itself. You can check for null if you just want to determine the presence of
an image.",
            "isDeprecated": false,
            "name": "back",
            "type": Object {
              "kind": "SCALAR",
              "name": "URLString",
              "ofType": null,
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "A list of images depicting the different sides and surfaces of a release’s
media and packaging.",
            "isDeprecated": false,
            "name": "images",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "LIST",
                "name": null,
                "ofType": Object {
                  "kind": "OBJECT",
                  "name": "CoverArtArchiveImage",
                  "ofType": null,
                },
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "Whether there is artwork present for this release.",
            "isDeprecated": false,
            "name": "artwork",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "Boolean",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "The number of artwork images present for this release.",
            "isDeprecated": false,
            "name": "count",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "Int",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "The particular release shown in the returned cover art.",
            "isDeprecated": false,
            "name": "release",
            "type": Object {
              "kind": "OBJECT",
              "name": "Release",
              "ofType": null,
            },
          },
        ],
        "inputFields": null,
        "interfaces": Array [],
        "kind": "OBJECT",
        "name": "CoverArtArchiveRelease",
        "possibleTypes": null,
      },
      Object {
        "description": "The image sizes that may be requested at the [Cover Art Archive](https://musicbrainz.org/doc/Cover_Art_Archive).",
        "enumValues": Array [
          Object {
            "deprecationReason": null,
            "description": "A maximum dimension of 250px.",
            "isDeprecated": false,
            "name": "SMALL",
          },
          Object {
            "deprecationReason": null,
            "description": "A maximum dimension of 500px.",
            "isDeprecated": false,
            "name": "LARGE",
          },
          Object {
            "deprecationReason": null,
            "description": "The image’s original dimensions, with no maximum.",
            "isDeprecated": false,
            "name": "FULL",
          },
        ],
        "fields": null,
        "inputFields": null,
        "interfaces": null,
        "kind": "ENUM",
        "name": "CoverArtArchiveImageSize",
        "possibleTypes": null,
      },
      Object {
        "description": "An individual piece of album artwork from the [Cover Art Archive](https://musicbrainz.org/doc/Cover_Art_Archive).",
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "The Internet Archive’s internal file ID for the image.",
            "isDeprecated": false,
            "name": "fileID",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "String",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "The URL at which the image can be found.",
            "isDeprecated": false,
            "name": "image",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "URLString",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "A set of thumbnails for the image.",
            "isDeprecated": false,
            "name": "thumbnails",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "OBJECT",
                "name": "CoverArtArchiveImageThumbnails",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "Whether this image depicts the “main front” of the release.",
            "isDeprecated": false,
            "name": "front",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "Boolean",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "Whether this image depicts the “main back” of the release.",
            "isDeprecated": false,
            "name": "back",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "SCALAR",
                "name": "Boolean",
                "ofType": null,
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "A list of [image types](https://musicbrainz.org/doc/Cover_Art/Types)
describing what part(s) of the release the image includes.",
            "isDeprecated": false,
            "name": "types",
            "type": Object {
              "kind": "NON_NULL",
              "name": null,
              "ofType": Object {
                "kind": "LIST",
                "name": null,
                "ofType": Object {
                  "kind": "SCALAR",
                  "name": "String",
                  "ofType": null,
                },
              },
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "The MusicBrainz edit ID.",
            "isDeprecated": false,
            "name": "edit",
            "type": Object {
              "kind": "SCALAR",
              "name": "Int",
              "ofType": null,
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "Whether the image was approved by the MusicBrainz edit system.",
            "isDeprecated": false,
            "name": "approved",
            "type": Object {
              "kind": "SCALAR",
              "name": "Boolean",
              "ofType": null,
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "A free-text comment left for the image.",
            "isDeprecated": false,
            "name": "comment",
            "type": Object {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
          },
        ],
        "inputFields": null,
        "interfaces": Array [],
        "kind": "OBJECT",
        "name": "CoverArtArchiveImage",
        "possibleTypes": null,
      },
      Object {
        "description": "URLs for thumbnails of different sizes for a particular piece of cover art.",
        "enumValues": null,
        "fields": Array [
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "The URL of a small version of the cover art, where the maximum dimension is
250px.",
            "isDeprecated": false,
            "name": "small",
            "type": Object {
              "kind": "SCALAR",
              "name": "URLString",
              "ofType": null,
            },
          },
          Object {
            "args": Array [],
            "deprecationReason": null,
            "description": "The URL of a large version of the cover art, where the maximum dimension is
500px.",
            "isDeprecated": false,
            "name": "large",
            "type": Object {
              "kind": "SCALAR",
              "name": "URLString",
              "ofType": null,
            },
          },
        ],
        "inputFields": null,
        "interfaces": Array [],
        "kind": "OBJECT",
        "name": "CoverArtArchiveImageThumbnails",
        "possibleTypes": null,
      },
    ],
  },
}
`;

exports[`run() with --help arg prints the help message 1`] = `
Object {
  "stderr": "",
  "stdout": "
  Usage: graphql-markdown [options] <schema>

  Output a Markdown document with rendered descriptions and links between types.
  The schema may be specified as:

    - a URL to the GraphQL endpoint (the introspection query will be run)
    - a GraphQL document containing the schema (.graphql or .gql)
    - a JSON document containing the schema (as returned by the introspection query)
    - an importable module with the schema as its default export (either an instance
      of GraphQLSchema or a JSON object)

  Options:

    --title <string>       Change the top heading title (default: 'Schema Types')
    --no-title             Do not print a default title
    --prologue <string>    Include custom Markdown after the title
    --epilogue <string>    Include custom Markdown after everything else
    --heading-level <num>  Heading level to begin at, useful if you are embedding the
                           output in a document with other sections (default: 1)
    --update-file <file>   Markdown document to update (between comment markers) or
                           create (if the file does not exist)
    --require <module>     If importing the schema from a module, require the specified
                           module first (useful for e.g. babel-register)
    --version              Print version and exit

",
}
`;

exports[`run() with no args prints the help message 1`] = `
Object {
  "stderr": "",
  "stdout": "
  Usage: graphql-markdown [options] <schema>

  Output a Markdown document with rendered descriptions and links between types.
  The schema may be specified as:

    - a URL to the GraphQL endpoint (the introspection query will be run)
    - a GraphQL document containing the schema (.graphql or .gql)
    - a JSON document containing the schema (as returned by the introspection query)
    - an importable module with the schema as its default export (either an instance
      of GraphQLSchema or a JSON object)

  Options:

    --title <string>       Change the top heading title (default: 'Schema Types')
    --no-title             Do not print a default title
    --prologue <string>    Include custom Markdown after the title
    --epilogue <string>    Include custom Markdown after everything else
    --heading-level <num>  Heading level to begin at, useful if you are embedding the
                           output in a document with other sections (default: 1)
    --update-file <file>   Markdown document to update (between comment markers) or
                           create (if the file does not exist)
    --require <module>     If importing the schema from a module, require the specified
                           module first (useful for e.g. babel-register)
    --version              Print version and exit

",
}
`;
